---
title: "Redrawing Enrichment plots for publication"
output: html_notebook
---

## Figure 2c

```{r, fig.width=3.3, fig.height=4.1}
library(dplyr)
library(tidyr)
library(ggplot2)
over_genome <- read.delim("../data/gene_sets/MSigDB_enhanced_genes_over_genome.tsv", stringsAsFactors = FALSE) %>%
  mutate(category=gsub("_", " ", category)) %>%
  mutate(category=gsub(".V1 (UP|DN)", "; \\1 Reg", category),
         category=gsub("HALLMARK", "", category),
         gs_cat = gsub("C6", "Oncogenic Signature", gs_cat),
         gs_cat = gsub("H", "Hallmarks", gs_cat))

over_genome %>%
  filter(gs_cat %in% c("Hallmarks", "Oncogenic Signature")) %>%
  mutate(category = substr(category, 1, 25)) %>%
  group_by(gs_cat) %>%
  arrange(desc(enrichment)) %>%
  dplyr::slice(1:20) %>%
    ungroup() %>%
  ggplot() +
   aes(x=reorder(category, enrichment), y=enrichment, col=-log10(padj)) +
   geom_point() +
   theme_bw(base_size = 8) +
   coord_flip() +
   xlab(NULL) +
   ylab("Fold Enrichment") +
   scale_color_continuous(guide=guide_colorbar(title.position = "top")) +
   theme(panel.spacing=grid::unit(0, "lines")) +
   facet_wrap(~gs_cat, ncol = 1, scale="free_y")-> g

ggsave("../Figures/Figure 2c - enrichments.pdf", g, width =3.3, height=4.1, dpi=300,useDingbats=FALSE)
print(g)

```

# Supplimentary figure 2a

```{r, fig.width=3.9, fig.height=2.4}

over_genome %>%
  filter(gs_cat %in% c("C2")) %>%
  mutate(category = substr(category, 1, 50),
         gs_subcat = ifelse(grepl("CP", gs_subcat), "Pathway", gs_subcat),
         category = str_to_title(category)) %>%
  arrange(desc(enrichment)) %>%
  filter(padj<0.01) %>%
  dplyr::slice(1:20) %>%
  ggplot() +
   aes(x=reorder(category, enrichment), y=enrichment, col=-log10(padj)) +
   geom_point() +
   theme_bw(base_size = 8) +
   coord_flip() +
   xlab(NULL) +
   ylab("Fold Enrichment") +
   scale_color_continuous(guide=guide_colorbar(title.position = "top"), limits=c(-1,NA)) +
   facet_grid(gs_subcat~., , scale="free_y", space="free" ) +
   theme(panel.spacing=grid::unit(0, "lines")) -> g
ggsave("../Figures/FigureS3a.png", g, width=3.9, height=2.4, dpi=300)
print(g)
```


## Suplimentary figure 3b

```{r, fig.width=3.3, fig.height=5.7}
library(stringr)
over_de <- read.delim("../data/gene_sets/enrichr_pathways_enhanced_genes_over_de_genes.tsv", 
                      stringsAsFactors = FALSE) %>%
  mutate(gs_cat = gsub("_", " ", gs_cat),
         category = str_to_sentence(category),
         category = gsub("mesc", "mESC", category),
         category = sub("^([A-Z][a-z]{1,2}[0-9]* )", "\\U\\1", category, perl=TRUE),
         category = sub("([a-z]+[0-9]+[a-z]*)", "\\U\\1", category, perl=TRUE),
         category = sub("chip", "ChIP", category),
         category = gsub("[0-9]{4,}", "", category))
  
over_de %>%
  filter(gs_cat %in% c("ENCODE Histone Modifications 2015",
                       "ChEA 2016",
                       "ESCAPE",
                       "ARCHS4 Tissues")) %>%
  mutate( gs_cat = sub("20[0-9]{2}", "", gs_cat)) %>%
  arrange(desc(enrichment)) %>%
  group_by(category) %>%
  dplyr::slice(1) %>%
  ungroup() %>%
  group_by(gs_cat) %>%
  dplyr::slice(1:20) %>%
  ungroup() %>%
  mutate(category = substr(category, 1, 30)) %>%
  ggplot() + aes(x=reorder(category, enrichment), y=enrichment, col=-log10(FDR)) +
   geom_point() +
   theme_bw(base_size = 8) +
   coord_flip() +
   xlab(NULL) +
   ylab("Fold Enrichment") +
   scale_color_continuous(guide=guide_colorbar(title.position = "top"), limits=c(-1,NA)) +
   facet_grid(gs_cat~., , scale="free_y", space="free" ) +
   theme(panel.spacing=grid::unit(0, "lines")) -> g

ggsave("../Figures/FigureS3b.pdf", g, width=3.3, height=5.7, useDingbats=FALSE)
  print(g)

```

## Figure S3c

```{r, fig.width = 3.2, fig.height = 5.2}
cell_type <- read.delim("../data/gene_sets/enrichr_pathways_enhanced_subtype_genes_over_de_genes.tsv") %>%
  mutate(gs_cat = gsub("_", " ", gs_cat),
         category = str_to_sentence(category),
         category = sub("Gtex-[^ ]+", "", category, perl=TRUE),
         category = gsub("mesc", "mESC", category),
         category = sub("^([A-Z][a-z]{1,2}[0-9]* )", "\\U\\1", category, perl=TRUE),
         category = sub("([a-z]+[0-9]+[a-z]*)", "\\U\\1", category, perl=TRUE),
         category = sub("chip", "ChIP", category),
         category = gsub("[0-9]{4,}", "", category),
         )

cell_type %>%
  filter( (gs_cat ==  "ARCHS4 Tissues" & subtype %in% c("HD", "MMSET")) | (grepl ("GTEx", gs_cat) & subtype == "MMSET")) %>%
  mutate(gs_cat=sub("Tissue Sample Gene Expression Profiles ", "", gs_cat)) %>%
  arrange(desc(OR)) %>%
  group_by(subtype, category) %>%
  dplyr::slice(1) %>%
  ungroup() %>%
  arrange(desc(OR)) %>%
  group_by(subtype, gs_cat) %>%
  dplyr::slice(1:20) %>%
  ungroup() %>%
  mutate(category = formatC(category, width=30, format="s")) %>%
  ggplot() + aes(x=reorder(paste0(subtype,category),  OR), y=OR, col=-log10(FDR)) +
   geom_point() +
   theme_bw(base_size = 8) +
   coord_flip() +
   xlab(NULL) +
   ylab("Odds Ratio") +
   scale_color_continuous(guide=guide_colorbar(title.position = "top"), limits=c(-1,NA)) +
   scale_y_continuous(breaks=c(2,2.5, 3, 3.5)) +
   scale_x_discrete(labels = function(x) sub("^(HD|MMSET)", "",x)) + 
   facet_grid(gs_cat + subtype ~ ., , scale="free_y", space="free" ) +
   theme(panel.spacing=grid::unit(0, "lines")) -> g

ggsave("../Figures/figureS3c.pdf", g, width = 3.2, height=5.2, useDingbats=FALSE)
print(g)
cell_type %>%
  filter( (gs_cat ==  "ARCHS4 Tissues" & subtype %in% c("HD", "MMSET")) | (grepl ("GTEx", gs_cat) & subtype == "MMSET")) %>%
  mutate(gs_cat=sub("Tissue Sample Gene Expression Profiles ", "", gs_cat)) %>%
  arrange(desc(OR)) %>%
  group_by(subtype, category) %>%
  dplyr::slice(1) %>%
  ungroup() %>% 
  filter(subtype=="MMSET", gs_cat=="GTEx up")
```

Add a new chunk by clicking the *Insert Chunk* button on the toolbar or by pressing *Ctrl+Alt+I*.

When you save the notebook, an HTML file containing the code and output will be saved alongside it (click the *Preview* button or press *Ctrl+Shift+K* to preview the HTML file).

The preview shows you a rendered HTML copy of the contents of the editor. Consequently, unlike *Knit*, *Preview* does not run any R code chunks. Instead, the output of the chunk when it was last run in the editor is displayed.
